### ISSUE TITLE
Fine-Grained Named Entity Recognition

### ISSUE THREAD

**Issue by alanakbik** [flairNLP/flair#1491]
The paper [Fine-Grained Named Entity Recognition](http://xiaoling.github.io/pubs/ling-aaai12.pdf) describes a dataset for NER with 112 different classes (called "FIGER"). This much richer set of classes than the standard 4-class CoNLL datasets might be useful for many users. Also, it would be interesting to see how well Flair does on this task and how it compares to published numbers. 

The authors make the dataset available [here](https://github.com/xiaoling/figer#training-data). 

Task: 

- [ ] Train fine-grained NER models for English with this data and report evaluation numbers
- [ ] Add best model for fine-grained NER as an included model to the `SequenceTagger`
- [ ] Create corpus object for FIGER in `flair.datasets` to auto-download this corpus so that other users can run experiments with this data

**Comment by alejandrojcastaneira** [comment]
I'm would be gladly wiling to help with by running the experiments in this dataset as I am interested in fine-grained entity recognition.

**Comment by alanakbik** [comment]
Hello @alejandrojcastaneira sorry for the late response (I somehow thought I had already replied)! Yes, we'd appreciate your help - we haven't yet started working on this, so a good starting point would be to check out the FIGER dataset and write a data loader for Flair. Let me know if you're still interested and if you have any questions!

**Comment by alejandrojcastaneira** [comment]
hi @alanakbik yes I am still interested, I'm checking the repository now and trying to build a corpus loader for this dataset once I have it ready, how can I share it?

**Comment by alanakbik** [comment]
@alejandrojcastaneira great! 

The easiest would be to open a pull request with the new dataset, then we can go over it. BTW the structure of `datasets` has changed so that each type of dataset is placed in a separate submodule. In your case, the data would go into `datasets/sequence_labeling.py` which is the submodule for all sequence labeling datasets.

**Comment by alejandrojcastaneira** [comment]
Hello @alanakbik an update and consultation. 

After extracting and preprocessing the dataset  I ended up with two sets:

A training set with around 29M tokens, ~1M sentences,  ~1.8M entity mentions and ~4M labels, average of 2,13 labels per entity, given that in the data generated using the _Distant Supervision's method_  a single mention could have multiple labels (max founded 8).

And a test set with 10008 tokens, 434 sentences, 562 mentions and 771 labels, this set was manually annotated, it's good to noticed that 217 of this labels corresponds to the type /person.

In the _FIGER_ hierchacy 2 levels of granularity were defined and 112 different entity types. 
I'm not sure how to define the schema to handle multiple labels for this dataset, should we built multiple sentences using all possible labels and permutations?  And should we train two models, one for the fist level in the hierarchy and one for the second one?  

Your advice is highly appreciated! and best regards

**Comment by alanakbik** [comment]
Hello @alejandrojcastaneira thanks for the update! I did not realize that entities had multiple annotations - could you post a few examples of this? When there are multiple annotations are they all correct or are most of them wrong and only one correct? Do you see a way of getting to a single annotation, maybe by only using the more coarse-grained hierarchy?

**Comment by alejandrojcastaneira** [comment]
Eg. training sample:

Fidel Castro and Che Guevara depart from Tuxpan , Veracruz , Mexico , enroute to Santiago de Cuba aboard the yacht Granma with 82 men .

entity: Fidel Castro , labels: ['/person/athlete', '/person/actor', '/person/politician', '/person', '/person/soldier', '/person/author']
entity: Che Guevara , labels: ['/person', '/organization/company', '/person/actor', '/person/author', '/person/doctor', '/person/artist', '/person/politician', '/person/soldier']
entity: Tuxpan , Veracruz , labels: ['/location/city', '/location']
entity: Mexico , labels: ['/location', '/person/artist', '/language', '/location/country']
entity: Santiago de Cuba , labels: ['/location', '/location/city']
entity: Granma , labels: ['/building', '/product/ship', '/location']

The distant supervision technique that was used to automatically build the training corpus assigned multiple labels to the hyperlinks in Wikipedia using their Freebases tags and then mapped those to the FIGER types, the main issue it's that all those types could be correct in different contexts, unfortunately to get the correct label depends of the context which it's still a subject of study of many papers.

A simple workaround it's to build all the possible sentences given the label combinations and train with this data, hopefully expecting that the model learns to differentiate contexts given the massive size of the annotation' corpus, but this training dataset could be relative huge. 

Also we could have a dataset for the coarse grained classes only, and other for the fine grained classes, although still we would have to find an agreement with the multiple labels.

**Comment by alanakbik** [comment]
Incredible, I did not realize that data quality was this problematic. Even "Mexico" is labeled as a "language" and an "artist" which is crazy. In this case, maybe FIGER is not the best dataset to train a fine-grained NER system.

There is a [recent paper](https://openreview.net/forum?id=HylHE-9p6m) by @abhipec that discusses the issues of distant supervision and presents an improved approach. Best part is that it is open source and they even [release all datasets here](https://github.com/abhipec/HAnDS/tree/master/datasets). Maybe one of these datasets could be used? 

@abhipec interesting work and thanks for making everything available - do you have a recommendation on which dataset we should use to train a fine-grained NER tagger?

**Comment by abhipec** [comment]
@alanakbik and @alejandrojcastaneira It is great that you are working on adding fine-grained entity recognition models on FLAIR. 

P1: [Fine-Grained Named Entity Recognition](http://xiaoling.github.io/pubs/ling-aaai12.pdf)
P2: [Fine-grained Entity Recognition with Reduced False Negatives and Large Type Coverage](https://openreview.net/pdf?id=HylHE-9p6m)

I would like to mention that in both P1 and P2, there are two models trained. One for the entity detection and other for the entity classification. In P1, the CoNLL dataset is used for entity detection model training, and the FIGER dataset is used for entity classification model training. In P2, the Wiki-FbF dataset is used for both entity detection and entity classification model training. To the best of my knowledge, the primary reason for the two-step approach is because sequence labeling is difficult in the multi-label setting. 

In terms of datasets, in P2, we have compared both FIGER and Wiki-FbF datasets on both the entity detection and classification tasks. In both these tasks, models trained on Wiki-FbF works better. This dataset also has less noise when compared with the FIGER dataset. So I will definitely recommend the Wiki-FbF dataset. 

All the code and datasets used in P2, are public at [https://github.com/abhipec/HAnDS](https://github.com/abhipec/HAnDS). I will be happy to answer any further queries.

Datasets:
FIGER: https://drive.google.com/open?id=0B52yRXcdpG6MMnRNV3dTdGdYQ2M
Wiki-FbF: https://drive.google.com/open?id=1WnV5gdichlXt9MNJAnL3bqtRx18W0ehd

**Comment by alanakbik** [comment]
Hello @abhipec thanks for the explanation! I looked a bit at Wiki-FbF and it does seem much less noisy than FIGER. 

A question about the labeling: In the following sentence from Wiki-FbF "_[...] a term used in the United Kingdom and Ireland to denote [...]_" there are several entities annotated, among them: 
```json
"links":[
      {
         "link":"United_Kingdom",
         "end":9,
         "name":"United Kingdom",
         "start":7,
         "labels":[
            "/location/country",
            "/organization",
            "/location",
            "/organization/government_agency"         
]      
},
      {
         "link":"Republic_of_Ireland",
         "end":11,
         "name":"Ireland",
         "start":10,
         "labels":[
            "/location/country",
            "/location"         
]      
```

So, while "Republic of Ireland" is top type `location`, "United Kingdom" is of multiple top types: `location` and `organization`. Do you then train and evaluate the taggers to predict all types?

Would it be possible (and make sense) to find a way so that each entity can be of only one top type? At least in the above example it seems that `location` is more fitting here than `organization`.

**Comment by abhipec** [comment]
Hi @alanakbik,

This issue is prevalent in all the fine-grained entity recognition datasets as the labels are obtained from a knowledge base, thus the labels in the datasets are context-independent. That is, the labels does not consider the sentence context into account. 

One way of addressing this issue is by some heuristic pruning. However, in the literature, there are several models which during training, tries to find out the best possible contextual label and outputs only single label or labels that correspond to a single path in the label hierarchy. Thus for these models, for the same entity mention occurring in different sentences, the labels will be different based on the entity mention's surrounding words.

In our experiments (related to the Wiki-FbF dataset) we used our own model published earlier at [EACL 2017](https://arxiv.org/abs/1702.06709). However, there are several other models proposed based on the idea of context-sensitive labels. The [latest paper](https://arxiv.org/abs/2004.02286)  was just uploaded few days ago on arXiv and is accepted at ACL 2020.

**Comment by alanakbik** [comment]
Hello @abhipec thanks for the clarification and the paper links - I did not realize that this issue affects all fine-grained NER datasets. This might be an interesting research direction to produce fine-grained NER datasets with "disambiguated" entity labels, so that there is less label noise. 

For our current sequence labeling implementation, the data is a bit problematic since we currently require one label per entity. We would need to make bigger changes to add fgNER to Flair (for instance by reimplementing your model). @alejandrojcastaneira still interested? ;)

**Comment by alejandrojcastaneira** [comment]
Hello @alanakbik  and @abhipec 

I agree that incorporating a model of this type to flair would dedicate a certain amount of time, between development and testing and some restructuring may be necessary, so it would be a long-term project. 

However in the [HAnDS](https://openreview.net/pdf?id=HylHE-9p6m) paper if I understood correctly the _**fgner**_ task it's divided in two steps: first a detector for the entities in a coarse granularity using a _LSTM-CNN-CRF_ model, and in a second step they get the fine granularities using the model proposed in their [EACL 2017](https://arxiv.org/pdf/1702.06709.pdf) paper.

As the model used  in the first step is similar to the model's type used in flair which requires only one type of label per entity. I would like to know @abhipec how you obtained and which fragment of _Wiki-FbF_ dataset to use for being able to train the first stage in the pipeline? 

I was observing in the repository and contains a directory named _datasets.tar.gz_ with the description: "Different samples of training datasets used in the paper").

**Comment by abhipec** [comment]
Hi @alejandrojcastaneira 

You are right. We used a standard sequence labeling algorithm (a combination of LSTM, CNN, and CRF) to identify entity mentions. So basically, it is two label sequence labeling tasks, entity, and non-entity labels.

The Wiki-FbF dataset has around 31 million sentences. We took a random sample of around 2 million sentences and trained the model on this sampled data. 

Multiple samples were taken to check for variations in samples. We observed less than 1% variation in results for different samples.

The result reported in the paper are on "RS_fner_3" sample, it is available in the file 'sampled_datasets.tar.gz'
